### PR TITLE
Allowing user to opt out of capturing pipeline data in image metadata

### PR DESCRIPTION
**Pull request by rakol-ms**
dockertask_no_pii

### CODE REVIEW

File: Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson [hunk 1]
```diff
@@ -20,6 +20,8 @@
   "loc.input.help.tags": "A list of tags in separate lines. These tags are used in build, push and buildAndPush commands. Ex:<br><br>beta1.1<br>latest",
   "loc.input.label.arguments": "Arguments",
   "loc.input.help.arguments": "Docker command options. Ex:<br> For build command,<br>--build-arg HTTP_PROXY=http://10.20.30.2:1234 --quiet",
+  "loc.input.label.hidePII": "Hide identifiable information from image metadata",
+  "loc.input.help.hidePII": "Prevent publishing identifiable information to image metadata store",
```
Comment (azooinmyluggage) [inline Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson]: By default pipeline data like source branch name, build id are added which helps with trace-ability. For example you can inspect an image to find out which pipeline built the image. You can opt out of this default behavior by using this input.

File: Tasks/DockerV2/task.json [hunk 2]
```diff
@@ -108,6 +108,14 @@
             "label": "Arguments",
             "groupName": "commands",
             "helpMarkDown": "Docker command options. Ex:<br> For build command,<br>--build-arg HTTP_PROXY=http://10.20.30.2:1234 --quiet"
+        },
+        {
+            "name": "hidePII",
```
Comment (azooinmyluggage) [inline Tasks/DockerV2/task.json]: addPipelineData

File: Tasks/DockerV2/task.json [hunk 3]
```diff
@@ -108,6 +108,14 @@
...
+            "type": "boolean",
+            "label": "Hide identifiable information from image metadata",
```
Comment (azooinmyluggage) [inline Tasks/DockerV2/task.json]: Add Pipeline metadata to image(s)

File: Tasks/DockerV2/task.json [hunk 4]
```diff
@@ -108,6 +108,14 @@
...
+            "groupName": "commands",
+            "defaultValue": false,
+            "helpMarkDown": "Prevent publishing identifiable information to image metadata store"
```
Comment (azooinmyluggage) [inline Tasks/DockerV2/task.json]: By default pipeline data like source branch name, build id are added which helps with trace-ability. For example you can inspect an image to find out which pipeline built the image. You can opt out of this default behavior by using this input.

**Comment by rakol-ms** [comment]
Task after changes:


docker_task

File: Tasks/Common/docker-common/pipelineutils.ts [hunk 5]
```diff
@@ -19,25 +19,31 @@ function addLabel(hostName: string, labelName: string, variableName: string, lab
...
+        addLabel(hostName, "system.teamproject", "SYSTEM_TEAMPROJECT", labels);
+        addLabel(hostName, "build.repository.name", "BUILD_REPOSITORY_NAME", labels);
+    }
 }
 
-function addBuildLabels(hostName: string, labels: string[]): void {
-    addLabel(hostName, "build.repository.uri", "BUILD_REPOSITORY_URI", labels);
-    addLabel(hostName, "build.sourcebranchname", "BUILD_SOURCEBRANCHNAME", labels);
+function addBuildLabels(hostName: string, labels: string[], addPipelineData?: boolean): void {
     addLabel(hostName, "build.sourceversion", "BUILD_SOURCEVERSION", labels);
```
Comment (ds-ms) [inline Tasks/Common/docker-common/pipelineutils.ts]: Any reason why this isn't inside the `if..else` block?

File: Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson [hunk 6]
```diff
@@ -20,6 +20,8 @@
   "loc.input.help.tags": "A list of tags in separate lines. These tags are used in build, push and buildAndPush commands. Ex:<br><br>beta1.1<br>latest",
   "loc.input.label.arguments": "Arguments",
   "loc.input.help.arguments": "Docker command options. Ex:<br> For build command,<br>--build-arg HTTP_PROXY=http://10.20.30.2:1234 --quiet",
+  "loc.input.label.addPipelineData": "Add Pipeline metadata to image(s)",
+  "loc.input.help.addPipelineData": "By default pipeline data like source branch name, build id are added which helps with trace-ability. For example you can inspect an image to find out which pipeline built the image. You can opt out of this default behavior by using this input.",
```
Comment (ds-ms) [inline Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson]: traceability is a single word.

Comment (ds-ms) [reply Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson]: Why not something like `When this is enabled, additional pipeline data would be added to your image; to improve traceability.`?

File: Tasks/DockerV2/Tests/L0.ts [hunk 7]
```diff
@@ -522,4 +535,49 @@ describe("DockerV2 Suite", function () {
...
+        console.log("\n");
+
+        setEnvironmentVariables();
+        process.env['SYSTEM_HOSTTYPE'] = 'build';
+        const labels = pipelineutils.getDefaultLabels(false); // addPipelineData is set to false
+
+        // update the label count in assert when newer labels are added
+        assert.equal(labels.length, 2, "Only selected labels are returned when addPipelineData is false");
+
+        delete process.env['SYSTEM_HOSTTYPE'];
```
Comment (ds-ms) [inline Tasks/DockerV2/Tests/L0.ts]: Why don't you add it to the `after` function?

File: Tasks/DockerV2/Tests/TestSetup.ts [hunk 8]
```diff
@@ -23,6 +23,7 @@ tr.setInput('Dockerfile', process.env[shared.TestEnvVars.dockerFile] || DefaultD
 tr.setInput('buildContext', process.env[shared.TestEnvVars.buildContext] || DefaultBuildContext);
 tr.setInput('tags', process.env[shared.TestEnvVars.tags] || "11");
 tr.setInput('arguments', process.env[shared.TestEnvVars.arguments] || "");
+tr.setInput ('addPipelineData', process.env[shared.TestEnvVars.addPipelineData] || "true");
```
Comment (ds-ms) [inline Tasks/DockerV2/Tests/TestSetup.ts]: Add a false test

File: Tasks/Common/docker-common/pipelineutils.ts (hunk 5 above)
Comment (rakol-ms) [reply Tasks/Common/docker-common/pipelineutils.ts]: It is that we would be adding few labels by default and would not add labels with PII if user opts out

File: Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson (hunk 6 above)
Comment (rakol-ms) [reply Tasks/DockerV2/Strings/resources.resjson/en-US/resources.resjson]: Changed the word `traceability`.

File: Tasks/DockerV2/Tests/TestSetup.ts (hunk 8 above)
Comment (rakol-ms) [reply Tasks/DockerV2/Tests/TestSetup.ts]: Done

File: Tasks/DockerV2/Tests/L0.ts [hunk 9]
```diff
@@ -522,4 +536,51 @@ describe("DockerV2 Suite", function () {
...
+        console.log("\n");
+
+        setEnvironmentVariables();
+        process.env['SYSTEM_HOSTTYPE'] = 'build';
+        const labels = pipelineutils.getDefaultLabels(true);
+
+        // update the label count in assert when newer labels are added
+        assert.equal(labels.length, 9, "All labels are returned by default");
+
+        delete process.env['SYSTEM_HOSTTYPE'];
```
Comment (ds-ms) [inline Tasks/DockerV2/Tests/L0.ts]: Remove this

File: Tasks/DockerV2/Tests/L0.ts [hunk 10]
```diff
@@ -522,4 +536,51 @@ describe("DockerV2 Suite", function () {
...
+        process.env[shared.TestEnvVars.command] = shared.CommandTypes.build;
+        process.env[shared.TestEnvVars.addPipelineData] = "false";
+        let tr : ttm.MockTestRunner = new ttm.MockTestRunner(tp);
+        tr.run();
+
+        assert(tr.invokedToolCount == 1, 'should have invoked tool one time. actual: ' + tr.invokedToolCount);
+        assert(tr.stderr.length == 0 || tr.errorIssues.length, 'should not have written to stderr');
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.stdout.indexOf(`[command]docker build -f ${shared.formatPath("a/w/Dockerfile")} ${shared.DockerCommandArgs.BuildLabelsWithAddPipelineFalse} -t testuser/testrepo:11 ${shared.formatPath("a/w")}`) != -1, "docker build should run with expected arguments");
+        console.log(tr.stderr);
```
Comment (ds-ms) [inline Tasks/DockerV2/Tests/L0.ts]: This line will always print empty, you can remove this.

**Pull request merged**